In [1]:
import sqlite3
import pandas as pd

In [2]:
pd.options.display.max_columns = None
#pd.options.display.max_rows = None

In [3]:
conn = sqlite3.connect('parch-and-posey.db')

In [4]:
cursor = conn.cursor()
cursor.execute('''
select * from sqlite_master where type = "table";
''')
columns = [col[0] for col in cursor.description]
data = cursor.fetchall()
cursor.close()

In [5]:
pd.DataFrame(data, columns=columns)

type        name    tbl_name  rootpage  \
0  table  web_events  web_events         2   
1  table  sales_reps  sales_reps        92   
2  table      region      region        93   
3  table      orders      orders        94   
4  table    accounts    accounts       221   

                                                 sql  
0  CREATE TABLE web_events (\tid integer,\taccoun...  
1  CREATE TABLE sales_reps (\tid integer,\tname b...  
2   CREATE TABLE region (\tid integer,\tname bpchar)  
3  CREATE TABLE orders (\tid integer,\taccount_id...  
4  CREATE TABLE accounts (\tid integer,\tname bpc...

Task 1:

Use DISTINCT to test if there are any accounts associated with more than one region.

In [45]:
# strange approach

pd.read_sql_query(sql='''
SELECT MIN(r.name), MAX(r.name)
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id
JOIN region r
ON s.region_id = r.id
GROUP BY a.name
HAVING MIN(r.name) != MAX(r.name);
''', con=conn)

Empty DataFrame
Columns: [MIN(r.name), MAX(r.name)]
Index: []

In [38]:
# count approach

pd.read_sql_query(sql='''
SELECT a.name, COUNT (*) count
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id
JOIN region r
ON s.region_id = r.id
GROUP BY a.name
ORDER BY count DESC;
''', con=conn)

name  count
0                   eBay      1
1             Yum Brands      1
2                  Xerox      1
3            Xcel Energy      1
4    World Fuel Services      1
..                   ...    ...
346                  AIG      1
347                  AES      1
348                AECOM      1
349                  ADP      1
350                   3M      1

[351 rows x 2 columns]

In [39]:
# compare number of results of all combinations to distinct

pd.read_sql_query(sql='''
SELECT a.name, r.name
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id
JOIN region r
ON s.region_id = r.id;
''', con=conn)

name       name
0               Walmart  Northeast
1           Exxon Mobil  Northeast
2                 Apple  Northeast
3    Berkshire Hathaway  Northeast
4              McKesson  Northeast
..                  ...        ...
346                 KKR       West
347               Oneok       West
348      Newmont Mining       West
349                 PPL       West
350         SpartanNash       West

[351 rows x 2 columns]

In [40]:
# compare number of results of all combinations to distinct

pd.read_sql_query(sql='''
SELECT DISTINCT id, name FROM accounts;
''', con=conn)

id                name
0    1001             Walmart
1    1011         Exxon Mobil
2    1021               Apple
3    1031  Berkshire Hathaway
4    1041            McKesson
..    ...                 ...
346  4461                 KKR
347  4471               Oneok
348  4481      Newmont Mining
349  4491                 PPL
350  4501         SpartanNash

[351 rows x 2 columns]

Solution:

The below two queries have the same number of resulting rows (351), so we know that every account is associated with only one region. If each account was associated with more than one region, the first query should have returned more rows than the second query.

SELECT a.id as "account id", r.id as "region id", 
a.name as "account name", r.name as "region name"
FROM accounts a
JOIN sales_reps s
ON s.id = a.sales_rep_id
JOIN region r
ON r.id = s.region_id;
and

SELECT DISTINCT id, name
FROM accounts;

Task 2:

Have any sales reps worked on more than one account?

In [48]:
# strange approach

pd.read_sql_query(sql='''
SELECT MIN(a.name), MAX(a.name)
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id
GROUP BY s.name
HAVING MIN(a.name) != MAX(a.name);
''', con=conn)

MIN(a.name)                       MAX(a.name)
0                           Comcast                             FedEx
1                 Applied Materials                     W.W. Grainger
2                 AmerisourceBergen                     Cisco Systems
3                               CBS                            Viacom
4            Charter Communications                    Voya Financial
5                               AES                         Supervalu
6             Bank of America Corp.               Northwestern Mutual
7                           Danaher                  Northrop Grumman
8                      Altria Group                         Whirlpool
9               Abbott Laboratories                         Tech Data
10                            Aflac                             Xerox
11                          Centene                            Kohl's
12                         AutoZone                              eBay
13                  Cardinal Health                    Morgan Stanley
14                            Amgen                Southwest Airlines
15                       Health Net             WellCare Health Plans
16   Cognizant Technology Solutions                       Toys 'R' Us
17                              ADP               United States Steel
18                              AIG          Twenty-First Century Fox
19                            Avnet          Walgreens Boots Alliance
20             Baxter International                           Veritiv
21                           Disney                     Valero Energy
22                         Best Buy                           PepsiCo
23         Air Products & Chemicals  Thrivent Financial for Lutherans
24           Archer Daniels Midland       Philip Morris International
25  American Family Insurance Group                    SunTrust Banks
26                         Allstate         State Farm Insurance Cos.
27                           AbbVie                              USAA
28                              EMC                  US Foods Holding
29                           DuPont                       Wells Fargo
30                     Baker Hughes                        Yum Brands
31               Advance Auto Parts              United Natural Foods
32                Core-Mark Holding                       Xcel Energy
33                       CBRE Group                  Western Refining
34                            Apple               United Technologies
35                            AECOM                        PG&E Corp.
36                       Amazon.com                   Lockheed Martin
37                             AT&T               World Fuel Services
38                Colgate-Palmolive                   Western Digital
39                            Aetna                       Time Warner
40                           Pfizer                           Verizon
41          American Airlines Group                           Walmart
42                              CSX                Whole Foods Market
43                        Citigroup              Prudential Financial
44         Community Health Systems                      U.S. Bancorp
45                               3M                               TJX
46                      Caterpillar                            Target
47                     Estee Lauder                          WestRock
48                         Alphabet       United Continental Holdings
49             Ameriprise Financial                           Textron

In [49]:
# count approach

pd.read_sql_query(sql='''
SELECT s.name, COUNT(*) count
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id
GROUP BY s.name
ORDER BY count DESC;
''', con=conn)

name  count
0     Georgianna Chisholm     15
1           Vernita Plump     11
2          Micha Woodford     11
3     Maryanna Fiorentino     11
4             Maren Musto     11
5       Earlie Schleusner     11
6         Dorotha Seawell     11
7          Calvin Ollison     11
8               Saran Ram     10
9             Moon Torian     10
10           Hilma Busick     10
11           Brandie Riva     10
12        Arica Stoltzfus     10
13      Marquetta Laycock      9
14           Elwood Shutt      9
15            Dawna Agnew      9
16              Tia Amato      8
17          Julia Behrman      8
18           Delilah Krum      8
19          Soraya Fulton      7
20            Nelle Meaux      7
21        Michel Averette      7
22          Gianna Dossey      7
23          Elna Condello      7
24        Derrick Boggess      7
25           Cliff Meints      7
26        Charles Bidwell      7
27         Babette Soukup      7
28           Sibyl Lauria      6
29          Samuel Racine      6
30         Necole Victory      6
31           Eugena Esser      6
32            Elba Felder      6
33         Debroah Wardle      6
34  Sherlene Wetherington      5
35            Chau Rowles      5
36      Carletta Kosinski      5
37          Renetta Carew      4
38            Lavera Oles      4
39       Kathleen Lalonde      4
40            Julie Starr      4
41            Cara Clarke      4
42         Silvana Virden      3
43         Shawanda Selke      3
44            Retha Sears      3
45           Nakesha Renn      3
46      Ernestine Pickron      3
47         Cordell Rieder      3
48          Ayesha Monica      3
49        Akilah Drinkard      3

In [44]:
# compare number of results of all combinations to distinct

pd.read_sql_query(sql='''
SELECT *
FROM accounts a
JOIN sales_reps s
ON a.sales_rep_id = s.id;
''', con=conn)

id                name                    website        lat  \
0    1001             Walmart            www.walmart.com  40.238496   
1    1011         Exxon Mobil         www.exxonmobil.com  41.169156   
2    1021               Apple              www.apple.com  42.290495   
3    1031  Berkshire Hathaway  www.berkshirehathaway.com  40.949021   
4    1041            McKesson           www.mckesson.com  42.217093   
..    ...                 ...                        ...        ...   
346  4461                 KKR                www.kkr.com  45.545353   
347  4471               Oneok              www.oneok.com  45.513513   
348  4481      Newmont Mining            www.newmont.com  45.494117   
349  4491                 PPL             www.pplweb.com  45.491720   
350  4501         SpartanNash        www.spartannash.com  45.555651   

           long       primary_poc  sales_rep_id      id                 name  \
0    -75.103297       Tamara Tuma        321500  321500        Samuel Racine   
1    -73.849374      Sung Shields        321510  321510         Eugena Esser   
2    -76.084009        Jodee Lupo        321520  321520      Michel Averette   
3    -75.763898    Serafina Banda        321530  321530        Renetta Carew   
4    -75.284998    Angeles Crusoe        321540  321540          Cara Clarke   
..          ...               ...           ...     ...                  ...   
346 -122.655247       Buffy Azure        321970  321970  Georgianna Chisholm   
347 -122.681500   Esta Engelhardt        321960  321960  Maryanna Fiorentino   
348 -122.669460  Khadijah Riemann        321970  321970  Georgianna Chisholm   
349 -122.671880   Deanne Hertlein        321960  321960  Maryanna Fiorentino   
350 -122.657145      Jewell Likes        321970  321970  Georgianna Chisholm   

     region_id  
0            1  
1            1  
2            1  
3            1  
4            1  
..         ...  
346          4  
347          4  
348          4  
349          4  
350          4  

[351 rows x 10 columns]

In [43]:
# compare number of results of all combinations to distinct

pd.read_sql_query(sql='''
SELECT DISTINCT id, name
FROM sales_reps;
''', con=conn)

id                   name
0   321500          Samuel Racine
1   321510           Eugena Esser
2   321520        Michel Averette
3   321530          Renetta Carew
4   321540            Cara Clarke
5   321550            Lavera Oles
6   321560            Elba Felder
7   321570         Shawanda Selke
8   321580           Sibyl Lauria
9   321590         Necole Victory
10  321600      Ernestine Pickron
11  321610          Ayesha Monica
12  321620            Retha Sears
13  321630          Julia Behrman
14  321640              Tia Amato
15  321650        Akilah Drinkard
16  321660         Silvana Virden
17  321670           Nakesha Renn
18  321680          Elna Condello
19  321690          Gianna Dossey
20  321700         Debroah Wardle
21  321710  Sherlene Wetherington
22  321720            Chau Rowles
23  321730      Carletta Kosinski
24  321740        Charles Bidwell
25  321750           Cliff Meints
26  321760           Delilah Krum
27  321770       Kathleen Lalonde
28  321780            Julie Starr
29  321790         Cordell Rieder
30  321800      Earlie Schleusner
31  321810            Moon Torian
32  321820        Dorotha Seawell
33  321830            Maren Musto
34  321840          Vernita Plump
35  321850         Calvin Ollison
36  321860              Saran Ram
37  321870        Derrick Boggess
38  321880         Babette Soukup
39  321890            Nelle Meaux
40  321900          Soraya Fulton
41  321910           Brandie Riva
42  321920      Marquetta Laycock
43  321930           Hilma Busick
44  321940        Arica Stoltzfus
45  321950           Elwood Shutt
46  321960    Maryanna Fiorentino
47  321970    Georgianna Chisholm
48  321980         Micha Woodford
49  321990            Dawna Agnew

Solution:

Actually all of the sales reps have worked on more than one account. The fewest number of accounts any sales rep works on is 3. There are 50 sales reps, and they all have more than one account. Using DISTINCT in the second query assures that all of the sales reps are accounted for in the first query.

SELECT s.id, s.name, COUNT(*) num_accounts
FROM accounts a
JOIN sales_reps s
ON s.id = a.sales_rep_id
GROUP BY s.id, s.name
ORDER BY num_accounts;
and

SELECT DISTINCT id, name
FROM sales_reps;